src/fromager/tarballs.py

```diff
@@ -33,5 +45,18 @@ def tar_reproducible(tar: tarfile.TarFile, basedir: pathlib.Path) -> None:
             content.append(os.path.join(root, filename))
     content.sort()
 
+    def _tar_filter(tarinfo: tarfile.TarInfo) -> tarfile.TarInfo:
+        tarinfo = _tar_reset(tarinfo)
+        if not prefix:
+            return tarinfo
+        # Ensure that the paths in the tarfile are rooted at the prefix directory
+        cur_name = tarinfo.name
+        if cur_name[0] != os.sep:
+            cur_name = os.sep + cur_name
+        new_name = pathlib.Path(cur_name).relative_to(prefix)
+        tarinfo.name = str(new_name)
+        logger.debug("adding %s", tarinfo.name)
+        return tarinfo
+
     for fn in content:
-        tar.add(fn, filter=_tar_reset, recursive=False, arcname=fn)
+        tar.add(fn, filter=_tar_filter, recursive=False)
```

Comment:
This code passes your tests and is easier to understand (at least for me):

```suggestion
        arcname = fn if prefix is None else os.path.relpath(fn, prefix)
        tar.add(fn, filter=_tar_reset, recursive=False, arcname=arcname)
```